File: springboot/src/main/java/com/example/tickets/dto/CreateTicketRequest.java
package com.example.tickets.dto;

import com.example.tickets.model.Priority;

// Java 21 record — immutable data carrier
// Rust parallel: struct CreateTicketRequest { title: String, priority: Priority, description: Option<String>, assignee: Option<String> }
public record CreateTicketRequest(
        String title,
        Priority priority,
        String description,  // null → Option::None
        String assignee      // null → Option::None
) {}

File: rust/src/schemas/create_ticket_request.rs
use serde::{Deserialize};
use validator::{Validate, ValidationError};

use crate::models::{priority::Priority};

#[derive(Deserialize, Validate)]
pub struct CreateTicketRequest {
    #[validate(custom(function = "not_blank"))]
    pub title: String,
    pub priority: Priority, // That's a leak of our domain model, but let's keep it simple for now
    pub description: Option<String>,
    pub assignee: Option<String>,
}

fn not_blank(value: &str) -> Result<(), ValidationError> {
    match value.trim().is_empty() {
        true => Err(ValidationError::new("not_blank").with_message("Title must not be blank.".into())),
        false => Ok(()),
    }
}